### ISSUE TITLE
Concurrency in Go - Fundamentals

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4739]
Concurrency in Go - Fundamentals<br>
<br>
https://ift.tt/2G5uP2g<br>
<br>
<br>
<br>
<div><article><p>In this post we will delve into the fundamentals of Go’s concurrency features, the target audience are readers familiar with the language; however no prior knowledge of concurrency is required.</p> <p>In this post we will talk about:</p> <ul>
<li>brief discussion of Go’s concurrency features</li> <li>goroutines</li> <li>WaitGroup:s (from the <strong>sync</strong> package)</li> <li>Mutex (also from <strong>sync</strong> package)</li> <li>channels (buffered &amp; unbuffered)</li> </ul>
<p>There’s definitely a lot more important subjects to cover but these are the basics that are vital to grasp.</p> <h2>Concurrency in Go</h2> <p>Without going into too much theory I would say that goroutines are at the core of concurrency in Go. Goroutines are user-space threads which are similar to kernel threads managed by the OS but instead managed entirely by the Go runtime. The reason for this is that user-space threads managed by the Go runtime are lighter-weight and cheaper than kernel threads. Also smaller memory footprint: initial goroutine stack = 2kb, default thread stack = 8kb</p> <p>Go is pretty unique in that concurrency is a first-class citizen in the language, you don’t need to download 100 dependencies or learn a framework to be able to write effective concurrent programs; everything you need is included in the standard library.</p> <p>Let’s move on by exploring goroutines</p> <h2>Goroutines</h2> <p>Goroutines is a function that is running concurrently, just put the keyworkd <strong>go</strong> in front of a function to make it run concurrently, example:</p> <p>The only interesting part in this simple example is the order of the print’s, running this example will output:</p> <pre><code>$ go run main.go Init End hello concurrency </code></pre> <p>The printer function is not running synchronous since we put the <strong>go</strong> keyword in front of it. You’ve might noticed the time.Sleep at line 16, lets delete that line and run the program again! Now the output is the following:</p> <pre><code>$ go run main.go Init End </code></pre> <p>Huh? Our printer function did not print it’s message… If you are not familiar with concurrency before this might seem a bit odd, however we just need to understand how our programs code execution works.</p> <p>Go follows a model of concurrency called the <strong>fork-join model</strong>. This means that our program can split into it’s own execution branch to be run concurrently with its main branch. At some point in the future, the two branches of execution will be joined together again. Let’s try to visually understand what is happening here:</p> <p><img alt="fork join" src="https://rogerwelin.github.io/assets/images/flow.png"></p> <p>We can clearly see that at <em>go printer…</em> the code execution flow branches off into its side branch while the main branch continues to run. After we print <em>End</em> the main branch (the program) terminates so our side branch never gets time to execute. What we need to do here is to wait for the side branch to finish it’s execution, join it in the main branch and then we can continue. Let’s use <em>WaitGroup</em> from the sync package to help us with that.</p> <h2>WaitGroup</h2> <p>Think of WaitGroup as a mechanism to fire off concurrent operations, let them do work and then wait for them to finish. The usage of WaitGroup is primarily when you don’t really need to know the result (example, fire off lots of http POSTs) or you have some other means to collect the resource (for example with channels which I will cover further on). Let’s look at a trivial example below (I’ll comment each number in the code snippet below):</p> <ol>
<li>This is our string slice with greek philosophers that we will print concurrently</li> <li>Here we declare a WaitGroup called wg</li> <li>For each element in the greeks slice we make a call to <em>wg.Add(1)</em> - wg.Add() adds to the WaitGroup counter. We then call the printer func concurenctly passing a pointer of wg</li> <li>In the printer func we use <em>defer wg.Done()</em> meaning if the WaitGroup counter was 2 it’s now 1</li> <li>
<em>wg.Wait()</em> blocks execution of the main branch of our program until the counter reaches zero</li> </ol>
<p>Try to run the program a couple of times to see that the output will differ in order each time.</p> <h4><strong>Concurrency Gotcha #1</strong></h4> <p>Instead of calling a separate func in our example above and passing a pointer of the WaitGroup you can use an anonymous function in the range loop which is an common pattern, this brings us to a gotcha shown below. Can you spot the problem?</p> <p>The output of our program is the following</p> <pre><code>$ go run main.go Init democritus democritus democritus democritus democritus democritus End </code></pre> <p>Pretty strange huh? In all fairness it’s pretty tricky to spot the error; what happens here is that there’s a high probability that the loop may exit before the goroutine begins, this means that the variable <em>ph</em> falls out of scope. But we actually got something printed out, why is that? Well, the go runtime knows that the <em>ph</em> variable is still being held and therefore will grab the last reference of that which are on the heap so that the goroutine can continue to access it.</p> <p>A correct implementation would look like:</p> <p>The trick here is that we need to pass a copy of <em>ph</em> to our anonymous func and thus making sure the goroutine will work on the data from the iteration of the loop:</p> <p>Examples up to this point has been very hello world-ish so let’s make a program that actually does something remotely useful. The program below does the following things: concurrently http get’s and collect the status code in a map, we are using WaitGroup as before</p> <p>This looks pretty straight forward, ….. But actually there’s a problem with this implementation which brings us to another concurrency gotcha; this code introduces a data race.</p> <h4><strong>Concurrency Gotcha #2 - data race</strong></h4> <p>So what’s a data race? A data race occurs when two or more goroutines access the same variable concurrently and at least one of the accesses is a write, this could lead to memory corruptions and crashes. It also makes the code unpredictable and hard to debug (potentially we can get different results each time the code is run).</p> <p>But why did we get a data race? Simply put; maps are not thread-safe! To verify this we can use the built-in tool <em>-race</em> to test our code:</p> <pre><code>$ go run -race main.go ================== WARNING: DATA RACE Write at 0x00c000096cc0 by goroutine 9: runtime.mapassign_faststr() ... Previous write at 0x00c000096cc0 by goroutine 6: runtime.mapassign_faststr() ... Found 1 data race(s) exit status 66 </code></pre> <p>Ouch! There are several ways we can fix this, for example with <em>channels</em> which I will cover shortly. Another way to fix this would be if we could protect our map but still run things concurrently; for this we use something called <em>mutex</em></p> <h2>Mutex</h2> <p>If you have a background in C/C++ mutex are probably familiar to you and they are similarly implemented in Go. Mutex stands for “mutual exclusion” and is a way to guard critical sections in your code from concurrent operations. To fix our data race from previous example we just need to put Lock() before the map operation and then Unlock() after the map operation has completed. Example below:</p> <figure><pre><code><span> </span><span>var</span><span> </span><span>wg</span><span> </span><span>sync</span><span>.</span><span>WaitGroup</span><span> </span><span>var</span><span> </span><span>mu</span><span> </span><span>sync</span><span>.</span><span>Mutex</span><span> </span><span>for</span><span> </span><span>_</span><span>,</span><span> </span><span>url</span><span> </span><span>:=</span><span> </span><span>range</span><span> </span><span>urls</span><span> </span><span>{</span><span> </span><span>wg</span><span>.</span><span>Add</span><span>(</span><span>1</span><span>)</span><span> </span><span>go</span><span> </span><span>func</span><span>(</span><span>url</span><span> </span><span>string</span><span>)</span><span> </span><span>{</span><span> </span><span>defer</span><span> </span><span>wg</span><span>.</span><span>Done</span><span>()</span><span> </span><span>resp</span><span>,</span><span> </span><span>err</span><span> </span><span>:=</span><span> </span><span>http</span><span>.</span><span>Get</span><span>(</span><span>url</span><span>)</span><span> </span><span>if</span><span> </span><span>err</span><span> </span><span>!=</span><span> </span><span>nil</span><span> </span><span>{</span><span> </span><span>fmt</span><span>.</span><span>Errorf</span><span>(</span><span>"url was error: %v"</span><span>,</span><span> </span><span>err</span><span>)</span><span> </span><span>}</span><span> </span><span>mu</span><span>.</span><span>Lock</span><span>()</span><span> </span><span>respStatus</span><span>[</span><span>url</span><span>]</span><span> </span><span>=</span><span> </span><span>resp</span><span>.</span><span>StatusCode</span><span> </span><span>mu</span><span>.</span><span>Unlock</span><span>()</span><span> </span><span>}(</span><span>url</span><span>)</span><span> </span><span>}</span><span> </span><span>wg</span><span>.</span><span>Wait</span><span>()</span></code></pre></figure><p>Try to run this program with <code>go run -race main.go</code> and you will see that the data race warnings are gone!</p> <h2>Using Channels</h2> <p>You can think of channels as concurrency safe containers, you associate any data type with them (int, string, map, struct and so on). Channels are ideal when you want to communicate information between goroutines. This is how you declare a channel that will hold string</p> <figure><pre><code><span>stream</span><span> </span><span>=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>string</span><span>)</span></code></pre></figure><p>You can define if the channel can only read or only write using the arrow character:</p> <figure><pre><code><span>// channel can both send and receive</span><span> </span><span>stream</span><span> </span><span>=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>string</span><span>)</span><span> </span><span>// channel can only read</span><span> </span><span>stream</span><span> </span><span>=</span><span> </span><span>make</span><span>(</span><span>chan</span><span>&lt;-</span><span> </span><span>string</span><span>)</span><span> </span><span>// channel can only send</span><span> </span><span>stream</span><span> </span><span>=</span><span> </span><span>make</span><span>(</span><span>chan</span><span>-&gt;</span><span> </span><span>string</span><span>)</span></code></pre></figure><p>Channel comes in two flavors:</p> <ul>
<li>buffered channels</li> <li>unbuffered channels</li> </ul>
<p>Unbuffered channels are declared without a size (as example above) while buffered channels are declared with a fixed size, like so:</p> <figure><pre><code><span>// buffered channel of strings buffering up to 2 values</span><span> </span><span>stream</span><span> </span><span>=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>string</span><span>,</span><span> </span><span>2</span><span>)</span></code></pre></figure><p>The difference initially seems trivial but there is a important implication in how they operate;</p> <ul>
<li>
<strong>unbuffered channel</strong>: a send operation on an unbuffered channel blocks the sending goroutine until another goroutine executes a receive on the same channel, now both goroutines may continue</li> <li>
<strong>unbuffered channel</strong>: if the receive operation was triggered first, the receiving goroutine is blocked until another goroutine performs a send on the same channel</li> <li>
<strong>buffered channel</strong>: send operation inserts element in the buffer/queue while a receive operation removes the element from the channel. If the channel is at maximum capacity the send operation blocks until space is “freed” by another goroutine’s receive.</li> <li>
<strong>buffered channel</strong>: if the buffer/queue is empty a receive oepration blocks until a value is sent by another goroutine</li> </ul>
<p>Let’s try to visualize these differences with some code examples:</p> <p>From what we’ve just learned on an <em>unbuffered channel</em> we should do the send and receive operation in different goroutines to avoid a deadlock, below is a simple demonstration of how to do this:</p> <figure><pre><code><span>func</span><span> </span><span>main</span><span>()</span><span> </span><span>{</span><span> </span><span>message</span><span> </span><span>:=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>string</span><span>)</span><span> </span><span>// we need to execute the send operation in a separate goroutine</span><span> </span><span>// as it would otherwise block the main goroutine </span><span> </span><span>go</span><span> </span><span>func</span><span>()</span><span> </span><span>{</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping"</span><span> </span><span>}()</span><span> </span><span>// execution blocks here until we receive something on channel message</span><span> </span><span>msg</span><span> </span><span>:=</span><span> </span><span>&lt;-</span><span>message</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>msg</span><span>)</span><span> </span><span>}</span></code></pre></figure><p><em>Buffered channel</em> does not block except when the channel is full, so we can do the send and receive operation in the same goroutine. However this example sends 3 elements on a channel with capacity of 2, thus our program will result in an deadlock:</p> <figure><pre><code><span>func</span><span> </span><span>main</span><span>()</span><span> </span><span>{</span><span> </span><span>// channel with capacity of 2</span><span> </span><span>message</span><span> </span><span>:=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>string</span><span>,</span><span> </span><span>2</span><span>)</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping 1"</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping 2"</span><span> </span><span>// at third element our channel is full thus our</span><span> </span><span>// program will block here and terminate the program with an deadlock error</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping 3"</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>&lt;-</span><span>message</span><span>)</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>&lt;-</span><span>message</span><span>)</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>&lt;-</span><span>message</span><span>)</span><span> </span><span>}</span></code></pre></figure><p>This brings us to another concurrency gotcha: a deadlock!</p> <h4><strong>Concurrency Gotcha #3 - deadlock</strong></h4> <p>A deadlock manifests itself when a program with concurrent processes are waiting one one another. In this state, the program will never recover. The Go compiler will notice this error on compiletime and will give us a warning and refuse to compile the program. Trying to run this program above will result in the following error:</p> <div><div><pre><code><span>$ </span>go run main.go fatal error: all goroutines are asleep - deadlock! goroutine 1 <span>[</span>chan send]: main.main<span>()</span> .../main.go:13 +0x9b <span>exit </span>status 2 </code></pre></div></div> <p>To make this code work we must unblock send operation from another goroutine, in the example below we do the send operation in a separate goroutine, it will block at element 3, but because the receive operation executes in a another goroutine the code will unblock:</p> <figure><pre><code><span>func</span><span> </span><span>main</span><span>()</span><span> </span><span>{</span><span> </span><span>// channel with capacity of 2</span><span> </span><span>message</span><span> </span><span>:=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>string</span><span>,</span><span> </span><span>2</span><span>)</span><span> </span><span>go</span><span> </span><span>func</span><span>()</span><span> </span><span>{</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping 1"</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping 2"</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>"ping 3"</span><span> </span><span>}()</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>&lt;-</span><span>message</span><span>)</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>&lt;-</span><span>message</span><span>)</span><span> </span><span>fmt</span><span>.</span><span>Println</span><span>(</span><span>&lt;-</span><span>message</span><span>)</span><span> </span><span>}</span></code></pre></figure><h4><strong>Closing Channels</strong></h4> <p>A channel is not a resource like a file or tcp connection that needs to be closed after use to free up resources, they can be left open infinitely. However we have the choice of closing them. Doing so indicates that no more value will be sent on the channel. This is useful if we want to avoid a deadlock and move on with our execution logic or communicate with other channels to return.</p> <p>A contrived but telling example is shown below. Here we create a buffered channel of type <em>person</em>, after writing to the channel we just want to iterate over the message without receiving from it. In normal cases this would result in a deadlock but since we closed the channel we’ve indicated that no more values can be sent to the channel.</p> <figure><pre><code><span>type</span><span> </span><span>person</span><span> </span><span>struct</span><span> </span><span>{</span><span> </span><span>Name</span><span> </span><span>string</span><span> </span><span>Age</span><span> </span><span>int</span><span> </span><span>}</span><span> </span><span>func</span><span> </span><span>main</span><span>()</span><span> </span><span>{</span><span> </span><span>message</span><span> </span><span>:=</span><span> </span><span>make</span><span>(</span><span>chan</span><span> </span><span>person</span><span>,</span><span> </span><span>3</span><span>)</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>person</span><span>{</span><span>Name</span><span>:</span><span> </span><span>"ping 1"</span><span>,</span><span> </span><span>Age</span><span>:</span><span> </span><span>1</span><span>}</span><span> </span><span>message</span><span> </span><span>&lt;-</span><span> </span><span>person</span><span>{</span><span>Name</span><span>:</span><span> </span><span>"ping 2"</span><span>,</span><span> </span><span>Age</span><span>:</span><span> </span><span>2</span><span>}</span><span> </span><span>close</span><span>(</span><span>message</span><span>)</span><span> </span><span>for</span><span> </span><span>element</span><span> </span><span>:=</span><span> </span><span>range</span><span> </span><span>message</span><span> </span><span>{</span><span> </span><span>fmt</span><span>.</span><span>Printf</span><span>(</span><span>"Age plus one is: %d</span><span>\n</span><span>"</span><span>,</span><span> </span><span>element</span><span>.</span><span>Age</span><span>+</span><span>1</span><span>)</span><span> </span><span>}</span><span> </span><span>}</span></code></pre></figure><p>Given what we learned so far of channel, let’s try to fix our concurrent urlfetcher program using only channels.</p> <p><strong>in progress</strong></p> </article></div><br>
<br>
<br>
<br>
<br>
<br>
via Roger Welin <br>
<br>
July 11, 2019 at 11:41AM
